### PR TITLE
Fix edge cases where inexact conversions to UInt don't throw

### LINKED ISSUE #51063: Incorrect conversion to unisigned integer
```julia
julia> convert(UInt32, 4.2949673f9)
0xffffffff
```

`4.2949673f9` is one greater than the maximum value of a `UInt32` (`4.2949673f9 == typemax(UInt32)+1`) so an inexact error should be thrown here:

```julia
function (::Type{$Ti})(x::$Tf)
    if ($(Tf(typemin(Ti))) <= x <= $(Tf(typemax(Ti)))) && isinteger(x)
        return unsafe_trunc($Ti,x)
    else
        throw(InexactError($(Expr(:quote,Ti.name.name)), $Ti, x))
    end
end
```

However, the comparison that happens is `4.2949673f9 <= Float32(typemax(UInt32))` which is true even though `4.2949673f9 > typemax(UInt32)`.

### PR DESCRIPTION
**Pull request by LilithHafner**
Fixes #51063

### CODE REVIEW

File: base/float.jl [hunk 1]
```diff
@@ -939,7 +939,7 @@ for Ti in (Int8, Int16, Int32, Int64, Int128, UInt8, UInt16, UInt32, UInt64, UIn
                     end
                 end
                 function (::Type{$Ti})(x::$Tf)
-                    if ($(Tf(typemin(Ti))) <= x <= $(Tf(typemax(Ti)))) && isinteger(x)
+                    if ($(Tf(typemin(Ti))) <= x < $(Tf(typemax(Ti))+one(Tf))) && isinteger(x)
```
Comment (Keno) [inline base/float.jl:942]: I think this deserves a source comment about why this isn't equivalent to `<=`.

**Review by Keno — COMMENTED** [review]
Seems quite clever, but I am slightly concerned about corner cases, since this is generic code. For example, if there happens to be a combination of integer and floating point types, with `eps(Tf(typemax(Ti))) == 2.0`, I think this might still have the issue of potentially giving the wrong answer. Can we instead just directly compute whether `Tf(typemax(Ti))` is larger than `typemax(Ti)` and if so, use `prevfloat(Tf(typemax(Ti)))` here?

**Comment by vtjnash** [comment]
LLVM also now has a well-behaved intrinsic that may not require these complications anymore to avoid these problems and the added cost of working around the undefined behavior. Would this be helpful? https://llvm.org/docs/LangRef.html#llvm-fptoui-sat-intrinsic

**Comment by LilithHafner** [comment]
Our semantics here are to throw, so the saturating version would not help.
The overhead should be low as well, simply comparing the input with compile time constant floating point values and if is not between them, throwing.
The saturating version would be great for #51069, though.

**Comment by LilithHafner** [comment]
I plan to go ahead and merge this when tests pass because it is
a) blocking https://github.com/JuliaLang/julia/pull/50812
b) a bug fix
c) thoroughly tested
d) a 1 LOC change

I'm still happy to receive comments and perform revisions in a followup.

**Comment by Keno** [comment]
> Technically, this is not generic code, it's metaprogramming to create methods with concrete type signatures; `eps(Tf(typemax(Ti))) == 2.0` never holds.

Good enough for me ;)

**Pull request merged**